apps/meteor/server/services/video-conference/service.ts

```diff
@@ -0,0 +1,238 @@
+			`config.prejoinPageEnabled=false`,
+			`config.requireDisplayName=false`,
+			`config.callDisplayName="Direct Message"`,
+		];
+
+		console.log(user.username, options);
+		if (options.mic !== undefined) {
+			configs.push(`config.startWithAudioMuted=${options.mic ? 'false' : 'true'}`);
+		}
+		if (options.cam !== undefined) {
+			configs.push(`config.startWithVideoMuted=${options.cam ? 'false' : 'true'}`);
+		}
+
+		const configHash = configs.join('&');
+		const url = `${call.url}#${configHash}`;
+
+		return url;
+	}
+}
+
+export class VideoConfService extends ServiceClassInternal implements IVideoConfService {
+	protected name = 'video-conference';
+
+	private Messages: MessagesRaw;
+
+	private Users: UsersRaw;
+
+	private Rooms: RoomsRaw;
+
+	private VideoConference: VideoConferenceRaw;
+
+	constructor(db: Db) {
+		super();
+
+		this.Users = new UsersRaw(db.collection('users'));
+		this.VideoConference = new VideoConferenceRaw(db.collection('rocketchat_video_conference'));
+		this.Rooms = new RoomsRaw(db.collection('rocketchat_room'));
+		this.Messages = new MessagesRaw(db.collection('rocketchat_message'));
+	}
+
+	public async start(caller: IUser['_id'], rid: string): Promise<VideoConferenceInstructions> {
+		const room = await this.Rooms.findOneById<Pick<IRoom, '_id' | 't' | 'uids'>>(rid, { projection: { t: 1, uids: 1 } });
+
+		if (!room) {
+			throw new Error('invalid-room');
+		}
+
+		if (room.t === 'd' && room.uids && room.uids.length <= 2) {
+			return this.startDirect(caller, room);
+		}
+
+		throw new Error('Conferece calls have not been implemented yet.');
+	}
+
+	public async join(uid: IUser['_id'], callId: IVideoConference['_id'], options: VideoConferenceJoinOptions): Promise<string> {
+		const call = await this.VideoConference.findOneById(callId);
+		if (!call) {
+			throw new Error('invalid-call');
+		}
+
+		const user = await this.Users.findOneById<Pick<IUser, '_id' | 'username' | 'name'>>(uid, { projection: { name: 1, username: 1 } });
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		switch (call.type) {
+			case 'direct':
+				return this.joinDirect(call, user, options);
+		}
+
+		throw new Error('Conferece calls have not been implemented yet.');
+	}
+
+	public async cancel(uid: IUser['_id'], callId: IVideoConference['_id']): Promise<void> {
+		const call = await this.VideoConference.findOneById(callId);
+		if (!call || !isDirectVideoConference(call)) {
+			throw new Error('invalid-call');
+		}
+
+		if (call.status !== VideoConferenceStatus.CALLING || call.endedBy || call.endedAt) {
+			throw new Error('invalid-call-status');
+		}
+
+		const user = await this.Users.findOneById(uid);
+		if (!user) {
+			throw new Error('failed-to-load-own-data');
+		}
+
+		if (call.messages.started) {
+			await this.changeMessageType(call.messages.started, 'video-direct-missed');
+		}
+
+		this.VideoConference.setEndedById(call._id, { _id: user._id, name: user.name, username: user.username });
```

Comment:
it is not mandatory to `await` here, but I think we should, to make sure when this function returns, the data is correctly saved on DB
```suggestion
		await this.VideoConference.setEndedById(call._id, { _id: user._id, name: user.name, username: user.username });
```